src/lib/statistics/table_statistics.cpp

```diff
@@ -1,317 +1,96 @@
 #include "table_statistics.hpp"
 
-#include <sstream>
+#include <numeric>
+#include <thread>
 
-#include "all_parameter_variant.hpp"
-#include "all_type_variant.hpp"
-#include "base_column_statistics.hpp"
+#include "column_statistics.hpp"
+#include "resolve_type.hpp"
+#include "statistics/statistics_objects/abstract_histogram.hpp"
+#include "storage/table.hpp"
+#include "utils/assert.hpp"
 
 namespace opossum {
 
-TableStatistics::TableStatistics(const TableType table_type, const float row_count,
-                                 const std::vector<std::shared_ptr<const BaseColumnStatistics>>& column_statistics)
-    : _table_type(table_type), _row_count(row_count), _column_statistics(column_statistics) {}
+std::shared_ptr<TableStatistics> TableStatistics::from_table(const Table& table) {
+  std::vector<std::shared_ptr<BaseColumnStatistics>> column_statistics(table.column_count());
 
-TableType TableStatistics::table_type() const { return _table_type; }
+  const auto histogram_bin_count = std::min<size_t>(100, std::max<size_t>(5, table.row_count() / 2'000));
 
-float TableStatistics::row_count() const { return _row_count; }
+  auto next_column_id = std::atomic<size_t>{0u};
+  auto threads = std::vector<std::thread>{};
 
-uint64_t TableStatistics::approx_valid_row_count() const {
-  return static_cast<size_t>(row_count()) - _approx_invalid_row_count;
-}
+  for (auto thread_id = 0u;
+       thread_id < std::min(static_cast<uint>(table.column_count()), std::thread::hardware_concurrency() + 1);
+       ++thread_id) {
+    threads.emplace_back([&] {
+      while (true) {
```

Comment:
A thread with a `while (true)` loop first looks like it is meant to run for a long time. Why is this necessary here?

Reply:
This is 1:1 copied from your encoding threading which also performs per column work.

Reply:
Added
```

  /**
   * Paralelly create statistics objects for the Table's columns
   */
```

Reply:
How about this (looking at benchmark_table_encoder.cpp):

```
      while (next_chunk + 1 < table->chunk_count()) {
        auto my_chunk = next_chunk++;
```

That seems to make the intent much clearer. No idea how the `while(true)` made it in back then.

Reply:
that alone is not thread-safe, we still need the `if (my_chunk >= table->chunk_count()) return;` check. And then we might just as well make it "while(true)" because the loop condition you proposed would essentially be pointless?